src/lib/work_partitioner/work_partitioner.ml

```diff
@@ -1 +1,242 @@
+open Core_kernel
+
+open struct
+  module Work = Snark_work_lib
+end
+
 module Snark_worker_shared = Snark_worker_shared
+module Zkapp_command_job_pool =
+  Job_pool.Make (Work.Id.Single) (Pending_zkapp_command)
+module Sent_zkapp_job_pool =
+  Job_pool.Make (Work.Id.Sub_zkapp) (Work.Spec.Sub_zkapp.Stable.Latest)
+module Sent_single_job_pool =
+  Job_pool.Make (Work.Id.Single) (Work.Spec.Single.Stable.Latest)
+
+(* TODO: use proper cache DB later *)
+let _proof_cache_db = Proof_cache_tag.create_identity_db ()
+
+(* WARN: we're mixing ID for `pairing_pool` and `zkapp_command_jobs. *)
+type t =
+  { logger : Logger.t
+  ; transaction_snark : (module Transaction_snark.S)
+  ; id_generator : Id_generator.t
+  ; pairing_pool : (int64, Combining_result.t) Hashtbl.t
+        (** if one single work from underlying Work_selector is completed but
+           not the other. throw it here. *)
+  ; zkapp_command_jobs : Zkapp_command_job_pool.t
+  ; reassignment_timeout : Time.Span.t
+  ; zkapp_jobs_sent_by_partitioner : Sent_zkapp_job_pool.t
+  ; single_jobs_sent_by_partitioner : Sent_single_job_pool.t
+  ; mutable tmp_slot :
+      (Work.Spec.Single.t * Work.Id.Single.t * Mina_base.Sok_message.t) option
+        (** When receving a `Two works from the underlying Work_selector, store
+            one of them here, so we could issue them to another worker. *)
+  }
+
+let create ~(reassignment_timeout : Time.Span.t) ~(logger : Logger.t) : t =
+  let module M = Transaction_snark.Make (struct
+    let constraint_constants = Genesis_constants.Compiled.constraint_constants
+
+    let proof_level = Genesis_constants.Compiled.proof_level
+  end) in
+  { logger
+  ; transaction_snark = (module M)
+  ; id_generator = Id_generator.create ~logger
+  ; pairing_pool = Hashtbl.create (module Int64)
+  ; zkapp_command_jobs = Zkapp_command_job_pool.create ()
+  ; reassignment_timeout
+  ; zkapp_jobs_sent_by_partitioner = Sent_zkapp_job_pool.create ()
+  ; single_jobs_sent_by_partitioner = Sent_single_job_pool.create ()
+  ; tmp_slot = None
+  }
+
+let epoch_now () = Time.(now () |> to_span_since_epoch)
+
+let resissue_if_old ~reassignment_timeout
+    (job : _ Work.With_job_meta.Stable.Latest.t) =
+  let issued = Time.of_span_since_epoch job.issued_since_unix_epoch in
+  let delta = Time.(diff (now ()) issued) in
+  if Time.Span.( > ) delta reassignment_timeout then
+    `Stop_reschedule
+      Work.With_job_meta.Stable.Latest.
+        { job with issued_since_unix_epoch = epoch_now () }
+  else `Stop_keep
+
+(* NOTE: below are logics for work requesting *)
+let reissue_old_zkapp_job
+    ~partitioner:
+      ({ reassignment_timeout; zkapp_jobs_sent_by_partitioner; _ } : t) () :
+    Work.Spec.Partitioned.Stable.Latest.t option =
```

Comment:
IMplementation of the two functions below is almost identical.
Let's define a common helper which will take `~reassignment_timeout jobs_sent_by_partitioner` as parameters.

Then this and next functions will be trivial one-liners

Reply:
I'll also need to add module sigs.